csrc/scheduler/matmul_utils.cpp

```diff
@@ -314,144 +314,302 @@ std::string isMatmulFusionDefinitionSupported(
-  supported_vec_size.b = getMinVectorization(MatmulRole::INPUT_B);
-  // Currently we set epilogue to the max vectorization supported by all outputs
-  // and all "C" type input dtypes.
-  // See https://github.com/NVIDIA/Fuser/issues/2169
-  supported_vec_size.epilogue = 16l;
-  // We will write OUTPUT_D role tensors in the default stride order. So
-  // vectorization is based on the inner dimension
-  const auto d_it = roles_map.find(MatmulRole::OUTPUT_D);
-  NVF_ERROR(d_it != roles_map.end(), "Could not find any output D tensors");
-  for (TensorView* tv : d_it->second) {
-    const int64_t N =
-        runtime_info.expressionEvaluator()
-            .evaluate(TensorDomain::noReductions(tv->getRFactorDomain())
-                          .back()
-                          ->extent())
-            .as<int64_t>();
-    supported_vec_size.epilogue =
-        std::min(supported_vec_size.epilogue, 16l / dataTypeSize(tv->dtype()));
-    supported_vec_size.epilogue = std::min(
-        supported_vec_size.epilogue, scheduler_utils::maxVectorizationWidth(N));
   }
-  // For INPUT_C role tensors, we do not necessarily know which axis we would
-  // like to vectorize, so we set vectorization based on dtype instead here
-  // until a more complete analysis is implemented.
-  if (const auto c_it = roles_map.find(MatmulRole::INPUT_C);
-      c_it != roles_map.end()) {
-    for (TensorView* tv : c_it->second) {
-      supported_vec_size.epilogue = std::min(
-          supported_vec_size.epilogue, 16l / dataTypeSize(tv->dtype()));
+
+  // Inspect the allocation domain of an operand input TensorView to determine
+  // vectorization width.
+  //
+  // We canonicalize dimensions by reordering them with the given ordering
+  // before merging all dimensions that have the same role. For a given operand,
+  // this might mean that the inner-most dimension gets reordered to be outer,
+  // even if it has the same role as the innermost dimension in the canonical
+  // ordering.
+  int64_t operandVectorization(TensorView* tv) {
+    // Check data pointer alignment
+    int64_t vec_size = ptrAndDTypeVec(tv);
+    if (vec_size == 1l) {
+      return vec_size;
     }
+
+    // Find the inner-most non-batch role for this tensor, and collect all
+    // ValGroups in that role, in the canonical ordering.
+    std::optional<MatmulDomain> vec_dim_role = std::nullopt;
+    for (int64_t i = (int64_t)(tv->getMaybeAllocationDomain().size()) - 1;
+         i >= 0;
+         --i) {
+      IterDomain* id = tv->getMaybeAllocationDomain()[i];
+      if (id->isDeviceDim() || id->isReduction() || id->isBroadcast()) {
+        continue;
+      }
+
+      ValGroup g = exact_graph_.toGroup(id);
+      MatmulDomain dim_role = dimRole(g);
+      if (dim_role == MatmulDomain::Batch) {
+        // We cannot vectorize in batch dimensions
+        break;
+      }
+      if (!vec_dim_role.has_value()) {
+        vec_dim_role = dim_role;
```

Comment:
should we do a `break;` after this line?

Reply:
Yes!

Reply:
Yes.